plugins/circleci/src/circleci-config.ts

```diff
@@ -1,29 +1,178 @@
 import { Hook } from '@dotcom-tool-kit/types'
+import { automatedComment, CircleConfig, Job, JobConfig, Workflow } from '@dotcom-tool-kit/types/lib/circleci'
 import { semVerRegex } from '@dotcom-tool-kit/types/lib/npm'
-import { Workflow, JobConfig, CircleConfig, automatedComment, Job } from '@dotcom-tool-kit/types/lib/circleci'
+import type { CircleCIOptions } from '@dotcom-tool-kit/types/lib/schema/circleci'
+import { promises as fs } from 'fs'
+import isMatch from 'lodash/isMatch'
+import merge from 'lodash/merge'
+import mergeWith from 'lodash/mergeWith'
+import path from 'path'
+import type { PartialDeep } from 'type-fest'
+import YAML from 'yaml'
+
+const majorOrbVersion = '3'
 
-const majorOrbVersion = '2'
+export type CircleCIState = CircleConfig
+/**
+ * CircleCIStatePartial makes every property in the config object optional,
+ * including nested properties. This is particularly useful as we merge each
+ * hook's config into the larger state, and so each hook only needs to define
+ * the parts of the config they want to add to.
+ */
+export type CircleCIStatePartial = PartialDeep<CircleCIState>
 
-interface CircleCIState {
-  jobs: Record<string, JobConfig>
-  nightlyJobs: Record<string, Omit<JobConfig, 'filters'>>
-  runOnVersionTags: boolean
-  additionalFields: Record<string, unknown>
+// These boilerplate objects are (typically) needed for each job. They can be
+// spread into your custom config, and are automatically included when calling
+// generateSimpleJob.
+
+/**
+ * Every Tool Kit job, including jobs in the `nightly` workflow, uses the
+ * executor we define at the top of the CircleCI config, which specifies the
+ * version of Node to use.
+ */
+export const nightlyBoilerplate = {
+  executor: 'node'
+}
+/**
+ * tagFilter sets the regex for GitHub release tags: CircleCI will ignore jobs
+ * when doing a release if the filter isn't made explicit
+ */
+export const tagFilter = { filters: { tags: { only: `${semVerRegex}` } } }
+/**
+ * jobBoilerplate is the config needed for all Tool Kit jobs in the `tool-kit`
+ * workflow, and combines the `nightlyBoilerplate` and `tagFilter` objects.
+ */
+export const jobBoilerplate = {
+  ...nightlyBoilerplate,
+  ...tagFilter
 }
 
-const tagFilter = { filters: { tags: { only: `${semVerRegex}` } } }
+export interface JobGeneratorOptions {
+  name: string
+  /** whether to include in `nightly` workflow or just `tool-kit` */
+  addToNightly: boolean
+  requires: string[]
+  /** other fields to include in the job */
+  additionalFields?: JobConfig
+}
+
+/**
+ * `generateConfigWithJob` generates a single job, structured so that it will
+ * merge nicely with the rest of the config. This will include the `requires`
+ * parameter, as well as the boilerplate properties from `jobBoilerplate`, but
+ * any other options will need to be passed to `additionalFields`, such as
+ * `filters.branches`.
+ */
+export const generateConfigWithJob = (options: JobGeneratorOptions): CircleCIStatePartial => {
```

Comment:
**praise:** this is a great function! Tidy, readable, and a helpful level of abstraction! Love it! `:dancing-otter:` 😁